rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+  Therefore all dependencies on other packages must come from `pkg-fun.nix` arguments injected by `callPackage`.
+  This ensures that files in nixpkgs can be moved around without breaking this package.
+- Files outside a unit directory must not reference files inside that unit directory.
+  Therefore other packages can only depend on this package via `pkgs.${name}`.
+  This ensures that files within unit directories (except `pkg-fun.nix`) can be freely moved and changed without breaking any other packages.
+
+The only notable exception to this rule is the `pkgs/top-level/all-packages.nix` file which may reference the `pkg-fun.nix` file according to the next requirement.
+
+## Custom arguments
+
+If `pkgs/top-level/all-packages.nix` contains a definition for the attribute `${name}` and the unit directory `pkgs/unit/${shard}/${name}` exists, then the attribute value must be defined as `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix args`, where `args` may be a freely chosen expression.
+
+This ensures that even if a package initially doesn't require a custom `args`, if it later does, it doesn't have to be moved out of the `pkgs/unit` directory to pass custom arguments.
+
+## Examples
+[examples]: #examples
+
+To add a new package `pkgs.foobar` to nixpkgs, one only needs to create the file `pkgs/unit/fo/foobar/pkg-fun.nix`.
+No need to find an appropriate category nor to modify `pkgs/top-level/all-packages.nix` anymore.
+
+With many packages, the `pkgs/unit` directory may look like this:
+
+```
+pkgs
+└── unit
+   ├── _0
+   │  ├── _0verkill
+   │  └── _0x
+   ┊
+   ├── ch
+   │  ├── ChowPhaser
+   │  ├── CHOWTapeModel
+   │  ├── chroma
+   │  ┊
+   ┊
+   ├── t
+   │  └── t
+   ┊
+```
+
+# Interactions
+[interactions]: #interactions
+
+- `nix edit` and search.nixos.org are unaffected, since they rely on `meta.position` to get the file to edit, which still works
+- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
+- A commonly recommended way of building package directories in nixpkgs is to use `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/applications/misc/hello {}'`.
+  Since the path changes `pkg-fun.nix` is now used, this becomes like `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/unit/he/hello/pkg-fun.nix {}'`, which is harder for users.
+  However, calling a path like this is an anti-pattern anyway, because it doesn't use the correct nixpkgs version and it doesn't use the correct argument overrides.
+  The correct way of doing it was to add the package to `pkgs/top-level/all-packages.nix`, then calling `nix-build -A hello`.
+  This `nix-build -E` workaround is partially motivated by the difficulty of knowing the mapping from attributes to package paths, which is what this RFC improves upon.
+  By teaching users that `pkgs/unit/*/<name>` corresponds to `nix-build -A <name>`, the need for such `nix-build -E` workarounds should disappear.
+
+# Drawbacks
```

Comment:
It is no longer possible to pass custom arguments with `callPackage` directly. This is done often enough in `all-packages.nix` and often with good reason.

I know this is mentioned in the Alternatives section, but I'd say it's a genuine drawback. I think such a change needs to have the ambition of being a global solution for everything it applies to (i.e. direct derivation children of `pkgs`, leaving out stuff related to toolchain logic etc.) which it is not in this form.

Reply:
That is still possible, see the [Custom Arguments](https://github.com/nixpkgs-architecture/rfcs/blob/269ac4f669302c9b09075a40632b664158908a0b/rfcs/0140-simple-package-paths.md#custom-arguments) section. Is this what you mean?

Reply:
I'm also adding the drawback that it only applies to top-level packages, and only ones defined with `callPackage`.